### PR TITLE
feat: implemented cookies API

### PR DESCRIPTION
**Pull request by Farfurix**
### TODO
- [x] rebase
- [x] API (hammerhead) return values
- [x] isValidCookieToGet -> isValidCookieToGetOrDelete
- [x] ~~`document.cookie` check: **need to synchronize cookies**~~
- [x] organize `_enqueueGetCookies`, `_enqueueSetCookies`, `_enqueueDeleteCookies`
- [x] ~~ts-defs-src/test-api/test-controller.d.ts, src/test-run/commands/actions.d.ts `...cookies: (Partial<Cookie>`~~ Next PR
- [x] fix server tests
- [x] callsites assertion in functional tests
- [x] check all added errors in tests
- [x] vars and hardcoded names in templates
- [x] ~~fix ts-ignore and any cases~~ Next PR
- [x] new URL(urlArgumentValue); // eslint-disable-line no-new
- [x] remove unnecessary functional tests after implementing server error tests
- [x] add "get" cases (only name, etc) to deleteCookies, delete {domain} case
- [x] check multiple "delete all" -> "set cookies" (undefined name issue)
- [ ] check `domain`/`path` matching cases
- [ ] check isClient http property case for API (hammerhead)
- [ ] check cookie validation functions
- [ ] check all messages and definitions (test-controller.d.ts - done)
- [ ] release a new hammerhead version
- [ ] getUrlCookieArgumentValidationError - unsupported protocols
- [ ] empty array argument cases: nameValueObjects, urls, ...cookies

## Test version installing
~~[testcafe-1.17.2-alpha.2-cookies.zip](https://github.com/DevExpress/testcafe/files/7625950/testcafe-1.17.2-alpha.2-cookies.zip)~~ outdated
```
npm i https://github.com/DevExpress/testcafe/files/7625950/testcafe-1.17.2-alpha.2-cookies.zip
```
`testcafe-hammerhead` version for this PR: https://github.com/DevExpress/testcafe-hammerhead/pull/2707

## API methods
```ts
/**
 * Returns cookies associated with the specified cookie objects. If there are no parameters, this method returns all cookies.
 *
 * @param cookies - The cookie objects.
 */
getCookies(...cookies: any): Promise<any>;
/**
 * Returns cookies with the specified names and URL contexts.
 *
 * @param names - The cookie names.
 * @param urls - The URL.
 */
getCookies(names: string | string[], urls?: string | string[]): Promise<any>;
/**
 * Sets the specified cookie objects.
 *
 * @param cookies - The cookie objects.
 */
setCookies(...cookies: any): TestControllerPromise;
/**
 * Sets cookies' name-value pairs to the specified URL contexts.
 *
 * @param nameValueObjects - The cookies' name-value pairs.
 * @param url - The URL.
 */
setCookies(nameValueObjects: Record<string, string> | Record<string, string>[], url: string): TestControllerPromise;
/**
 * Deletes cookies. If there are no parameters, this method deletes all cookies.
 *
 * @param cookies - The cookies.
 */
deleteCookies(...cookies: any): TestControllerPromise;
/**
 * Deletes cookies associated with the URL contexts. If there are no parameters, this method deletes all cookies.
 *
 * @param names - The cookie names.
 * @param urls - The URL.
 */
deleteCookies(names: string | string[], urls?: string | string[]): TestControllerPromise;
```

## Simple test (set cookies)
```js
import { RequestLogger } from 'testcafe';

fixture `Cookies API`;

test('Should set cookies', async t => {
    const logger = RequestLogger(/google.com/, { logRequestHeaders: true });

    await t.addRequestHooks(logger);

    const cookiesThatShouldNotBeInRequestHeaders = [
        { name: 'apiCookie13', value: 'value13', domain: 'some-another-domain.com', path: '/' },
        { name: 'apiCookie14', value: 'value14', domain: 'some-another-domain.com', path: '/' },
    ];

    // Set cookies via t.setCookies(nameValueObjects, url)
    await t
        .setCookies([
                { apiCookie1: 'value1' }, { 'apiCookie2': 'value2' }, { 'apiCookie3': 'value3' },
                { 'apiCookie4': 'value4' }, { 'apiCookie5': 'value5' }, { 'apiCookie6': 'value6' },
                { 'apiCookie7': 'value7' }, { 'apiCookie8': 'value8' }, { 'apiCookie9': 'value9' },
                { 'apiCookie10': 'value10' }, { 'apiCookie11': 'value11' }, { 'apiCookie12': 'value12' },
            ],
            'http://google.com')
        .setCookies(cookiesThatShouldNotBeInRequestHeaders);

    const cookies = await t.getCookies();

    // Delete all cookies to set them (the `cookies` variable) in another way
    await t.deleteCookies();

    // Set the cookies via t.setCookies(...cookies)
    await t
        .setCookies(cookies[0])
        .setCookies([cookies[1], cookies[2]])
        .setCookies(cookies[3], cookies[4])
        .setCookies(cookies[5], cookies[6], cookies[7])
        .setCookies([cookies[8], cookies[9], cookies[10]], [cookies[11], cookies[12]], cookies[13]);

    await t.navigateTo('https://google.com');

    // Check cookies in request headers
    await t
        .expect(logger.requests[0].request.headers.cookie)
        .eql('apiCookie1=value1; apiCookie2=value2; apiCookie3=value3; apiCookie4=value4; ' +
            'apiCookie5=value5; apiCookie6=value6; apiCookie7=value7; apiCookie8=value8; ' +
            'apiCookie9=value9; apiCookie10=value10; apiCookie11=value11; apiCookie12=value12');
});

```
## More tests
[test/functional/fixtures/api/es-next/cookies/testcafe-fixtures/cookies-test.js](https://github.com/DevExpress/testcafe/pull/6707/files#diff-378bdcade3dd5934c0ca2ce5c10446f9af372e609e2e4fb20508c144279d153c): "Should get cookies", "Should delete cookies"

## Pre-Merge TODO
- [ ] Write tests for your proposed changes
- [ ] Make sure that existing tests do not fail

### CODE REVIEW

File: src/test-run/index.ts [hunk 1]
```diff
@@ -775,6 +781,168 @@ export default class TestRun extends AsyncEventEmitter {
...
+
+        if (cookies) {
+            for (const cookie of cookies) {
+                const domain = cookie.domain;
+                const path   = cookie.path;
+                const name   = cookie.name;
+
+                if (domain) {
+                    if (name)
+                        cookiesPromises.push(this.session.cookies.findCookieByApi(domain, path, name));
```
Comment (miherlosev) [inline src/test-run/index.ts:800]: Can we change signature of `session.cookies.findCookieByApi` function to allowing it asset an array of arguments?

**Review by miherlosev — CHANGES_REQUESTED** [review]

File: src/api/test-controller/index.js [hunk 2]
```diff
@@ -506,7 +669,7 @@ export default class TestController {
 
     _addRequestHooks$ (...hooks) {
         return this._enqueueTask('addRequestHooks', () => {
-            hooks = flatten(hooks);
+            hooks = flattenDeep(hooks);
```
[the remark this answers is not in the document]
Comment (Farfurix) [reply src/api/test-controller/index.js:671]: Previosly, we had [the `flattenDeep as flatten` import](https://github.com/DevExpress/testcafe/pull/6707/files#diff-7bef381dd93d2a93756020c28b7c0ace5f4f51c99e7731e37e55ac6f20fc7326L7). In my case I need the original `flatten` method (one level deep). In this PR I imported original methods ([flatten](https://lodash.com/docs/4.17.15#flatten), [flattenDeep](https://lodash.com/docs/4.17.15#flattenDeep)), so there is no difference in behavior for hooks.

File: src/test-run/commands/actions.d.ts [hunk 3]
```diff
@@ -228,3 +228,24 @@ export class ScrollIntoViewCommand extends ActionCommandBase {
     public selector: ExecuteClientFunctionCommand;
     public options: ActionOptions;
 }
+
+export class GetCookiesCommand extends ActionCommandBase {
+    public constructor(obj: object, testRun: TestRun, validateProperties: boolean);
+    public names: string | string[];
+    public urls: string | string[];
+    public cookies: any;
```
[the remark this answers is not in the document]
Comment (Farfurix) [reply src/test-run/commands/actions.d.ts:236]: Yes, it's already on the TODO list.

File: src/test-run/commands/validations/argument.js [hunk 4]
```diff
@@ -127,3 +140,124 @@ export function functionArgument (name, val) {
...
+            return cookieArgumentsLength === 1
+                ? new ActionCookieArgumentError(callsite)
+                : new ActionCookieArgumentsError(callsite, cookieArgumentIndex);
+        }
+    }
+
+    return null;
+}
+
+export function cookieArgumentsToGetOrDelete (callsite, cookieArguments) {
```
Comment (miherlosev) [inline src/test-run/commands/validations/argument.js:188]: Method names should start with the verb (`get`, `set`, `validate`, etc.)

File: ts-defs-src/test-api/test-controller.d.ts [hunk 5]
```diff
@@ -98,6 +98,19 @@ interface WindowFilterData {
 
 type ScrollPosition = 'top' | 'right' | 'bottom' | 'left' | 'topRight' | 'topLeft' | 'bottomRight' | 'bottomLeft' | 'center';
 
+interface Cookie {
```
Comment (miherlosev) [inline ts-defs-src/test-api/test-controller.d.ts:101]: Let's add a public type definition after the feature is completely implemented. Now, just remove it.

**Review by miherlosev — CHANGES_REQUESTED** [review]

**Pull request closed without merging**